typed/union_test.go

```diff
@@ -0,0 +1,224 @@
+      type:
+        scalar: string
+    - name: a
+      type:
+        scalar: numeric
+    - name: b
+      type:
+        scalar: numeric
+    - name: c
+      type:
+        scalar: numeric
+    - name: one
+      type:
+        scalar: numeric
+    - name: two
+      type:
+        scalar: numeric
+    - name: three
+      type:
+        scalar: numeric
+    unions:
+      - discriminator: discriminator
+        fields: [a, b, c]
+      - fields: [one, two, three]`)
+	if err != nil {
+		panic(err)
+	}
+	return parser.Type("union")
+}()
+
+func TestNormalizeUnions(t *testing.T) {
+	tests := []struct {
+		name string
+		old  typed.YAMLObject
+		new  typed.YAMLObject
+		out  typed.YAMLObject
+	}{
+		{
+			name: "nothing changed, nothing changes",
+			old:  `{"a": 1}`,
+			new:  `{"a": 1}`,
+			out:  `{"a": 1}`,
+		},
+		{
+			name: "proper union update, setting discriminator",
+			old:  `{"a": 1}`,
+			new:  `{"b": 1}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "proper union update, no discriminator",
+			old:  `{"one": 1}`,
+			new:  `{"two": 1}`,
+			out:  `{"two": 1}`,
+		},
+		{
+			name: "proper union update from not-set, setting discriminator",
+			old:  `{}`,
+			new:  `{"b": 1}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "proper union update from not-set, no discriminator",
+			old:  `{}`,
+			new:  `{"two": 1}`,
+			out:  `{"two": 1}`,
+		},
+		{
+			name: "remove union, with discriminator",
+			old:  `{"a": 1}`,
+			new:  `{}`,
+			out:  `{}`,
+		},
+		{
+			name: "remove union and discriminator",
+			old:  `{"a": 1, "discriminator": "a"}`,
+			new:  `{}`,
+			out:  `{}`,
+		},
+		{
+			name: "remove union, not discriminator",
+			old:  `{"a": 1, "discriminator": "a"}`,
+			new:  `{"discriminator": "a"}`,
+			out:  `{"discriminator": "a"}`,
+		},
+		{
+			name: "remove union, no discriminator",
+			old:  `{"two": 1}`,
+			new:  `{}`,
+			out:  `{}`,
+		},
+		{
+			name: "dumb client update, no discriminator",
+			old:  `{"one": 1}`,
+			new:  `{"one": 2, "two": 1}`,
+			out:  `{"two": 1}`,
+		},
+		{
+			name: "dumb client update, sets discriminator",
+			old:  `{"a": 1}`,
+			new:  `{"a": 2, "b": 1}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "dumb client doesn't update discriminator",
+			old:  `{"a": 1, "discriminator": "a"}`,
+			new:  `{"a": 2, "b": 1, "discriminator": "a"}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "client sends multiple with discriminator",
+			old:  `{}`,
+			new:  `{"a": 1, "b": 1, "discriminator": "a"}`,
+			out:  `{"a": 1, "discriminator": "a"}`,
+		},
+		{
+			name: "multi-discriminator at the same time",
+			old:  `{"one": 1, "a": 1}`,
+			new:  `{"one": 1, "two": 1, "a": 1, "b": 1}`,
+			out:  `{"two": 1, "b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "change discriminator, nothing else",
+			old:  `{"discriminator": "a"}`,
+			new:  `{"discriminator": "random"}`,
+			out:  `{"discriminator": "random"}`,
```

Comment:
random discriminator values should cause an error, don't you think?

Reply:
I think this is similar to the one below, which is to make sure changing a DeploymentStrategyType from RollingUpdate to Recreate clears the RollingUpdate field.
Recreate isn't a name of a field in the Union but it is an allowed value of DeploymentStrategyType. If it's not allowed as a DeploymentStrategyType, I think validation outside smd could handle that kind of thing.